x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportGetShutdownStatusAction.java

```diff
@@ -111,6 +150,95 @@ protected void masterOperation(
+        int totalRemainingShards = relocatingShards + startedShards + initializingShards;
+
+        // If there's relocating shards, or no shards on this node, we'll just use the number of shards left to move
+        if (relocatingShards > 0 || totalRemainingShards == 0) {
+            SingleNodeShutdownMetadata.Status shardStatus = totalRemainingShards == 0
+                ? SingleNodeShutdownMetadata.Status.COMPLETE
+                : SingleNodeShutdownMetadata.Status.IN_PROGRESS;
+            return new ShutdownShardMigrationStatus(shardStatus, totalRemainingShards);
+        } else if (initializingShards > 0 && relocatingShards == 0 && startedShards == 0) {
+            // If there's only initializing shards left, return now with a note that only initializing shards are left
+            return new ShutdownShardMigrationStatus(
+                SingleNodeShutdownMetadata.Status.IN_PROGRESS,
+                totalRemainingShards,
+                "all remaining shards are currently INITIALIZING and must finish before they can be moved off this node"
+            );
+        }
+
+        // If there's no relocating shards and shards still on this node, we need to figure out why
+        final RoutingAllocation allocation = new RoutingAllocation(
+            allocationDeciders,
+            currentState.getRoutingNodes(),
+            currentState,
+            clusterInfoService.getClusterInfo(),
+            snapshotsInfoService.snapshotShardSizes(),
+            System.nanoTime()
+        );
+        allocation.debugDecision(true);
+
+        // Explain shard allocations until we find one that can't move, then stop (as `findFirst` short-circuits)
+        final Optional<Tuple<ShardRouting, ShardAllocationDecision>> unmovableShard = currentState.getRoutingNodes()
+            .node(nodeId)
+            .shardsWithState(ShardRoutingState.STARTED)
+            .stream()
+            .map(shardRouting -> new Tuple<>(shardRouting, allocationService.explainShardAllocation(shardRouting, allocation)))
+            // Given that we're checking the status of a node that's shutting down, no shards should be allowed to remain
+            .filter(pair -> {
+                assert pair.v2().getMoveDecision().canRemain() == false
+                    : "shard [" + pair + "] can remain on node [" + nodeId + "], but that node is shutting down";
+                return pair.v2().getMoveDecision().canRemain() == false;
+            })
+            // It's okay if some are throttled, they'll move eventually
+            .filter(pair -> pair.v2().getMoveDecision().getAllocationDecision().equals(AllocationDecision.THROTTLED) == false)
+            // These shards will move as soon as possible
+            .filter(pair -> pair.v2().getMoveDecision().getAllocationDecision().equals(AllocationDecision.YES) == false)
+            .findFirst();
```

Comment:
If you don't want to use the allocation decision (from my comment above, because you don't use the `v2()` anywhere but for filtering), this could just be:

```java
.map(Tuple::v1)
.findFirst();
```

And then it only needs to be an `Optional<ShardRouting>` above

Reply:
Good call, thanks!